inmem/session_store.go

```diff
@@ -0,0 +1,82 @@
+package inmem
+
+import (
+	"sync"
+	"time"
+)
+
+type SessionStore struct {
+	data map[string]string
+
+	timers map[string]*time.Timer
+
+	mu sync.RWMutex
+}
+
+func NewSessionStore() *SessionStore {
+	return &SessionStore{
+		data:   map[string]string{},
+		timers: map[string]*time.Timer{},
+	}
+}
+
+func (s *SessionStore) Set(key, val string, expireAt time.Time) error {
+	if !expireAt.IsZero() && expireAt.Before(time.Now()) {
+		// key is already expired. no problem
+		return nil
+	}
+
+	s.mu.Lock()
+	s.data[key] = val
+	s.mu.Unlock()
+
+	if !expireAt.IsZero() {
+		return s.ExpireAt(key, expireAt)
+	}
+	return nil
+}
+
+func (s *SessionStore) Get(key string) (string, error) {
+	s.mu.RLock()
+	defer s.mu.RUnlock()
+
+	return s.data[key], nil
+}
+
+func (s *SessionStore) Delete(key string) error {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	timer := s.timers[key]
+	if timer != nil {
+		timer.Stop()
+	}
+
+	delete(s.data, key)
+	delete(s.timers, key)
+	return nil
+}
+
+func (s *SessionStore) ExpireAt(key string, expireAt time.Time) error {
+	s.mu.Lock()
+
+	existingTimer, ok := s.timers[key]
+	if ok {
+		existingTimer.Stop()
```

Comment:
Food for though: https://golang.org/pkg/time/#Timer.Stop
I know this is mostly for internal / testing use (I assume anyway). The after func goroutine may have already entered here and be waiting on the Lock currently held. In which case `Delete(key)` will be called once the lock here is released, after this function returns.
In terms of practicality it probably doesn't matter too much, this would only happen given a new expiration was added around the same time it was due to expire and Delete doesn't explode if the key does not exist.
Just a little racey, but I can't imagine it being a big problem.
Could always just bottom out here and return nil or return an key expired error or something here when `!existingTime.Stop()`.

Reply:
I will add some data safety when the timer was already stopped.